### PR TITLE
Ignore/include properties in application.properties depending on test

### LINKED ISSUE #5: Ignore/include properties in application.properties depending on test scope
According @fbricon:

We need to consider the "test" flag of the source directory the application.properties is coming from. 

 * if no "test" flag, then key/value pairs from test dependencies should not be considered to be returned to the client (edited) 

 * if "test" flag is on, then key/value pairs from test dependencies should also be returned (edited) 

As a use case, imagine an application.properties in src/test/resources, that would be able to use a specific test database (from a quarkus extension available in the test scope), in order to override a production DB defined in src/main/resources/application.properties

### PR DESCRIPTION
**Pull request by angelozerr**
Ignore/include properties in application.properties depending on test

Fixes #5

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR gives the capability : 

 * to collect Quarkus properties only from src if application.properties is in src
 * to collect Quarkus properties from src&test if application.properties is in test
 * returns an empty array of Quarkus properties if application.properties is not in the classpath.

The QuarkusProjectInfo host a new information ClasspathKind to know if application.properties is in src, test our outside the classpath

This information could be used in the fiture to add a warning to the application.properties which are not in the classpath.

I set this PR as draft, because it mises the test but @fbricon if you wish you can play with it.

**Comment by angelozerr** [comment]
@fbricon I rebase the PR from master and there are tests in https://github.com/redhat-developer/quarkus-ls/pull/95/files#diff-62241c40b3f540ce53094c6b772edf4a

Please review it.

Please note that QuarkusProjectInfo returns the classpath kind (NONE, SRC, TEST). We could in the future add an error in application.properties root if it is not included in classpath (NONE)

**Comment by fbricon** [comment]
It doesn't seem to work as expected.

If I have src/main/resources/application.properties and src/test/resources/application.properties, both containing 

> quarkus.hibernate-orm.database.generation=create

Both files show the warning about the unknown property.

Now if I add 
```xml
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-hibernate-orm-panache</artifactId>
      <scope>test</scope>
    </dependency>
```

Both files' warnings disappear. Only the test one should be warning-free. And if I remove the test scope, then the test properties shows the warning. And then, trying once more I get different results, the main properties get the warning.

Expected behavior:
- dependency in compile scope : visible by both the main and test properties
- dependency in test scope : visible by both the test properties only

**Comment by angelozerr** [comment]
@fbricon my PR works if you add Java sources ConfigProperty or ConfigRoot in src/test/java. I have not tested your usecase but I don't if JDT Search is able to do that (using maven scope)?

**Comment by fbricon** [comment]
I still see erratic behavior even after applying #103

**Comment by fbricon** [comment]
The problem might be linked to detecting the scope change: when you change the scope, the  classpath didn't really change except for the test flag being toggled on the classpath entry. I'm not sure the classpath change detector handles that

File: quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/core/JDTQuarkusManager.java [hunk 1]
```diff
@@ -111,13 +121,14 @@ private JDTQuarkusManager() {
...
-
-		QuarkusProjectInfo info = new QuarkusProjectInfo();
-		info.setProjectURI(JDTQuarkusUtils.getProjectURI(javaProject));
-
+	public QuarkusProjectInfo getQuarkusProjectInfo(IJavaProject javaProject, QuarkusPropertiesScope propertiesScope,
+			DocumentationConverter converter, ClasspathKind classpathKind, IProgressMonitor monitor)
+			throws JavaModelException, CoreException {
+		QuarkusProjectInfo info = createInfo(javaProject, classpathKind);
+		if (classpathKind == ClasspathKind.NONE) {
+			info.setProperties(new ArrayList<>());
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/core/JDTQuarkusManager.java]: Collection.emptyList()

Comment (angelozerr) [reply quarkus.jdt/com.redhat.quarkus.jdt.core/src/main/java/com/redhat/quarkus/jdt/core/JDTQuarkusManager.java]: @fbricon updated

File: quarkus.jdt/com.redhat.quarkus.jdt.test/src/main/java/com/redhat/quarkus/jdt/core/JDTQuarkusManagerClassPathKindTest.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+				// Optional<String> name;
+				p(null, "greeting.name", "java.util.Optional", null, "/application-configuration-test/src/main/java",
+						"org.acme.config.GreetingResource#name", 0, null));
+
+		// in /java/main/test classpath-> N + M quarkus properties
+		IFile filefromTest = javaProject.getProject().getFile(new Path("src/test/resources/application.properties"));
+		QuarkusProjectInfo infoFromTest = JDTQuarkusManager.getInstance().getQuarkusProjectInfo(filefromTest,
+				QuarkusPropertiesScope.classpath, DocumentationConverter.DEFAULT_CONVERTER, new NullProgressMonitor());
+		Assert.assertEquals(ClasspathKind.TEST, infoFromTest.getClasspathKind());
+		assertProperties(infoFromTest, 69 /* properties from JAR */ + 3 /* properties from (src) Java sources */
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.test/src/main/java/com/redhat/quarkus/jdt/core/JDTQuarkusManagerClassPathKindTest.java]: should check properties from test jars are collected too

File: quarkus.jdt/com.redhat.quarkus.jdt.test/src/main/java/com/redhat/quarkus/jdt/core/JDTQuarkusManagerClassPathKindTest.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+		// in /java/main/test classpath-> N + M quarkus properties
+		File undertowJARFile = DependencyUtil.getArtifact("io.quarkus", "quarkus-undertow-deployment", "0.21.2", null);
+		Assert.assertNotNull("Test existing of quarkus-undertow-deployment*.jar", undertowJARFile);
+
+		IFile filefromTest = javaProject.getProject().getFile(new Path("src/test/resources/application.properties"));
+		QuarkusProjectInfo infoFromTest = JDTQuarkusManager.getInstance().getQuarkusProjectInfo(filefromTest,
+				QuarkusPropertiesScope.classpath, DocumentationConverter.DEFAULT_CONVERTER, new NullProgressMonitor());
+		Assert.assertEquals(ClasspathKind.TEST, infoFromTest.getClasspathKind());
+		assertProperties(infoFromTest, 63 /* properties from JAR */ + 22 /* properties from JAR (test) */ + 3 /*
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.test/src/main/java/com/redhat/quarkus/jdt/core/JDTQuarkusManagerClassPathKindTest.java]: empty lines

File: quarkus.jdt/com.redhat.quarkus.jdt.test/src/main/java/com/redhat/quarkus/jdt/core/JDTQuarkusManagerClassPathKindTest.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+						"org.acme.config.GreetingResource#suffix", 0, "!"),
+
+				// @ConfigProperty(name = "greeting.name")
+				// Optional<String> name;
+				p(null, "greeting.name", "java.util.Optional", null, "/application-configuration-test/src/main/java",
+						"org.acme.config.GreetingResource#name", 0, null));
+
+		// in /java/main/test classpath-> N + M quarkus properties
+		File undertowJARFile = DependencyUtil.getArtifact("io.quarkus", "quarkus-undertow-deployment", "0.21.2", null);
+		Assert.assertNotNull("Test existing of quarkus-undertow-deployment*.jar", undertowJARFile);
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.test/src/main/java/com/redhat/quarkus/jdt/core/JDTQuarkusManagerClassPathKindTest.java]: "quarkus-undertow-deployment*.jar is missing"

File: quarkus.jdt/com.redhat.quarkus.jdt.test/src/main/java/com/redhat/quarkus/jdt/core/JDTQuarkusManagerClassPathKindTest.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+		// not in classpath -> 0 quarkus properties
+		IFile fileFromNone = javaProject.getProject().getFile(new Path("application.properties"));
+		QuarkusProjectInfo infoFromNone = JDTQuarkusManager.getInstance().getQuarkusProjectInfo(fileFromNone,
+				QuarkusPropertiesScope.classpath, DocumentationConverter.DEFAULT_CONVERTER, new NullProgressMonitor());
+		Assert.assertEquals(ClasspathKind.NONE, infoFromNone.getClasspathKind());
+		Assert.assertEquals(0, infoFromNone.getProperties().size());
+
+		File resteasyJARFile = DependencyUtil.getArtifact("io.quarkus", "quarkus-resteasy-common-deployment", "0.21.2",
+				null);
+		Assert.assertNotNull("Test existing of quarkus-resteasy-common-deployment*.jar", resteasyJARFile);
```
Comment (fbricon) [inline quarkus.jdt/com.redhat.quarkus.jdt.test/src/main/java/com/redhat/quarkus/jdt/core/JDTQuarkusManagerClassPathKindTest.java]: "quarkus-resteasy-common-deployment*.jar is missing"

**Comment by angelozerr** [comment]
@fbricon anything else?

**Pull request merged**